modules/gmake/gmake.lua

```diff
@@ -0,0 +1,338 @@
+		:select({ projects = name })
+		:fromScopes(wks.root)
+		:withInheritance()
+	)
+
+	prj.root = wks.root
+	prj.workspace = wks
+
+	local name = os.matchFiles(path.join(prj.location, 'Makefile')) and prj.name .. '.mak' or 'Makefile'
+	local prjPath = path.join(prj.location, name)
+
+	prj.exportPath = prjPath
+	prj.baseDirectory = path.getDirectory(prj.exportPath)
+	prj.uuid = prj.uuid or os.uuid(prj.name)
+
+	-- TODO: Compiler/Linker from DOM
+	prj.compiler = gmake.toolsets.compilers.gcc
+	prj.linker = gmake.toolsets.linkers.gcc
+
+	prj.configs = prj:fetchConfigs(gmake.fetchProjectConfig)
+
+	table.sort(prj.configs, function(cfg0, cfg1)
+		return (cfg0.gmake_identifier:lower() < cfg1.gmake_identifier:lower())
+	end)
+
+	return prj
+end
+
+
+---
+-- Fetch the settings for a specific workspace-level configuration.
+--
+-- @param wks
+--	The `dom.Workspace` instance which contains the target configuration.
+-- @param build
+--	The target build configuration name, eg. 'Debug'
+-- @param platform
+--	The target platform name.
+-- @returns
+--	A `dom.Config`, with additional Visual Studio specific values.
+---
+function gmake.fetchWorkspaceConfig(wks, build, platform)
+	local cfg = gmake.fetchConfig(wks
+		:selectAny({ configurations = build, platforms = platform })
+		:fromScopes(wks.root)
+		:withInheritance()
+	)
+
+	cfg.root = wks.root
+	cfg.workspace = wks
+
+	return cfg
+end
+
+
+---
+-- Fetch the settings for a specific project-level configuration.
+--
+-- @param wks
+--	The `dom.Project` instance which contains the target configuration.
+-- @param build
+--	The target build configuration name, eg. 'Debug'
+-- @param platform
+--	The target platform name.
+-- @returns
+--	A `dom.Config`, with additional Visual Studio specific values.
+---
+function gmake.fetchProjectConfig(prj, build, platform)
+	local cfg = gmake.fetchConfig(prj
+		:selectAny({ configurations = build, platforms = platform })
+		:fromScopes(prj.root, prj.workspace)
+	)
+
+	cfg.root = prj.root
+	cfg.workspace = prj.workspace
+	cfg.project = prj
+
+	-- TODO: Compiler/Linker from DOM
+	cfg.compiler = gmake.toolsets.compilers.gcc
+	cfg.linker = gmake.toolsets.linkers.gcc
+
+	-- Configurations inherit most values from project, but files should be kept separated
+	cfg.files = cfg:withoutInheritance().files
+
+	return cfg
+end
+
+
+---
+-- Fetch the settings for a specific file-level configuration.
+--
+-- @param cfg
+--	The `dom.Config` instance which contains the target file settings. May be a
+--	workspace or project configuration.
+-- @param file
+--	The path of the file for which settings should be fetched.
+-- @returns
+--	A `dom.Config`, with additional Visual Studio specific values.
+---
+function gmake.fetchFileConfig(cfg, file)
+	local fileCfg = dom.Config.new(cfg
+		:select({ files = file })
+		:fromScopes(cfg.root, cfg.workspace, cfg.project)
+	)
+
+	fileCfg.file = file
+	fileCfg.project = cfg.project
+	fileCfg.gmake_build = cfg.gmake_build
```

Comment:
I'd say set a `fileCfg.config` here instead, and then reference these projects like `fileCfg.config.project` and `fileCfg.config.gmake_build` instead? Seems cleaner than copying? But not a big deal if these are the only fields that get duplicated.

Reply:
Not in use yet, so easy to change